### PR TITLE
[REF] Move Email, Address, etc. is_primary handling on delete to a hook (dev/core#2757) 

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Part of ongoing refactoring work for dev/core#2757

Before
----------------------------------------
Post-delete logic to handle `is_primary` in a delegated function.

After
----------------------------------------
Post-delete logic moved to a hook listener.

### CODE REVIEW

**Comment by colemanw** [comment]
@eileenmcnaughton there are existing tests covering this behavior.

**Comment by eileenmcnaughton** [comment]
@colemanw do you know where this is specifically covered - I actually struggled to find it - although https://github.com/civicrm/civicrm-core/pull/21402 should fish it out

**Comment by eileenmcnaughton** [comment]
@colemanw I dont' think anything tests delete handling specifically (in that there are no asserts ) - but can you add a line into this test between the 2 deletes & it that is covered I'm good with this

```
  public function testCreateDuplicateLocationTypes() {
    $address1 = $this->callAPISuccess('address', 'create', $this->_params);
    $address2 = $this->callAPISuccess('address', 'create', [
      'location_type_id' => $this->_locationTypeID,
      'street_address' => '1600 Pensilvania Avenue',
      'city' => 'Washington DC',
      'is_primary' => 0,
      'is_billing' => 0,
      'contact_id' => $this->_contactID,
    ]);
    $check = $this->callAPISuccess('address', 'getcount', [
      'contact_id' => $this->_contactID,
      'location_type_id' => $this->_locationTypeID,
    ]);
    $this->assertEquals(2, $check);
    $this->callAPISuccess('address', 'delete', ['id' => $address1['id']]);
    $this->callAPISuccess('address', 'delete', ['id' => $address2['id']]);
  }
```

**Comment by colemanw** [comment]
@eileenmcnaughton I've added  an APIv4 test

**Comment by eileenmcnaughton** [comment]
@colemanw style

File: CRM/Core/DAO.php [hunk 1]
```diff
@@ -962,7 +962,9 @@ public static function deleteRecord(array $record) {
     CRM_Utils_Hook::pre('delete', $entityName, $record['id'], $record);
     $instance = new $className();
     $instance->id = $record['id'];
-    if (!$instance->delete()) {
+    // Load complete object for the sake of post-hooks
+    $instance->find(TRUE);
```
Comment (eileenmcnaughton) [inline CRM/Core/DAO.php:966]: @colemanw could you add a comment here explaining what you explained on chat - ie that we can't expect post hooks to do this load themselve in the case of delete as there is nothing left to load. With that added this is MOP

**Comment by colemanw** [comment]
Comments added.

**Pull request merged**